### PR TITLE
Add direct commit mode to ps-singer-tap

### PR DESCRIPTION
**Pull request by Phanatic**
Introduce a `commit` mode so that `ps-singer-tap` can publish records to [Stitch's Import API](https://www.stitchdata.com/docs/developers/import-api/), instead of stdout.

``` bash
 go run cmd/singer-tap/main.go -h  2>&1 | grep -1 "sync mode"
  -batch-size int
    	(sync mode only) size of each batch sent to Singer (default 9000)
  -catalog string
    	(sync mode only) path to a catalog file for this tap
  -commit
    	(sync mode only) Run this tap in commit mode, sends rows to Stitch Import API
  -singer-api-token string
    	(sync mode only) API Token to authenticate with Singer
  -singer-api-url string
    	(sync mode only) API Url for Singer (default "https://api.stitchdata.com")
  -state string
    	(sync mode only) path to state file for this configuration
  -state-directory string
    	(sync mode only) Directory to save any received state
```

### Before

``` bash
ps-singer-tap --config sources/largedb/source.json --catalog sources/largedb/name_data.json | ps-http-tap --api-token $STITCH_API_TOKEN
PlanetScale Tap : INFO : PlanetScale Singer Tap : version [""], commit [""], published on [""]
PlanetScale Tap : INFO : Syncing records for PlanetScale database : large-import-56
PlanetScale Tap : INFO : Stream "name_data" will be synced incrementally
PlanetScale Tap : INFO : known position is CgEtEg9sYXJnZS1pbXBvcnQtNTY=
PlanetScale Tap : INFO : syncing rows from stream "name_data" from shard "-" at position []
PlanetScale Tap : INFO : [name_data shard : -] peeking to see if there's any new rows
PlanetScale Tap : INFO : [name_data shard : -] syncing rows with cursor [shard:"-"  keyspace:"large-import-56"]
PlanetScale Tap : INFO : [name_data shard : -] latest database position is [MySQL56/094dfc93-c7f4-11ed-85a0-1af10d8c33d9:1-3676,0952eb04-c7f4-11ed-9488-9a951227f355:1-1175,8bf226de-ae66-11ed-8a7f-8e9fa0edf0b2:1-403,8c4d148f-ae66-11ed-a066-bed353af0f03:1-852,8d08e564-d884-11ed-b9fc-92807b25c579:1-2223,a6f10029-d7c4-11ed-80c4-3aff882caf60:1-1607,a9a3719d-d7c4-11ed-beb2-4a7ff2069748:1-12476,f4c25a4b-a7d9-11ed-86d1-8e2dfefa1b03:1-1759,f5c41ba5-a7d9-11ed-8902-369ef0b7d59c:1-4380]
PlanetScale Tap : INFO : [name_data shard : -] Finished reading all rows for table [name_data]
HTTP Tap : INFO : saving state to path : state/state-1684947055433.json
HTTP Tap : INFO : saving state to path : state/state-1684947055433.json
HTTP Tap : INFO : saving state to path : state/state-1684947055433.json
HTTP Tap : INFO : Published [1701] records for stream "name_data"
```

☝🏾 what's happening here is that the `ps-singer-tap` is writing to `stdout`, and `ps-http-tap` is reading from `stdin`.
Fun fact : `stdin` and `stdout` are both buffered, and the buffer size is set to the kernel page size. You can find the buffersize for both by running `pagesize` on your machine. 

1.  This input buffer, combined with the buffer in the `ps-http-tap` that is used to read from `stdio`, give us **many possible points of failures for losing records when writing to stitch**. 
2. In the above output, the **state is written before the records are sent** to Stitch Import API. 
3. This gives the customer a false confidence that all rows were written to stitch, when in reality we could have missed a few. 
And when a customer uses the state file to continue, they move on and never get the rows missed from the last sync session. 
4. When writing to stdout, if the `ps-singer-tap` pauses because the server disconnected or we got timed out, the write to stdout is paused, which can cause the pipe to break and shut down the sync process. 

Given a two CLI approach, some of these issues are inevitable and make debugging dataflow issues difficult for us. 

### After

``` bash
$ ps-singer-tap --config sources/largedb/source.json --catalog sources/largedb/name_data.json --commit --singer-api-token $STITCH_API_TOKEN --state-directory state/

PlanetScale Tap : INFO : PlanetScale Singer Tap : version [""], commit [""], published on [""]
PlanetScale Tap : INFO : Syncing records for PlanetScale database : large-import-56
{"type":"SCHEMA","stream":"name_data","tap_stream_id":"large-import-56:name_data","schema":{"type":["null","object"],"additionalProperties":false,"properties":{"name":{"type":["null","string"]},"name_type":{"type":["null","string"]}}},"table-name":"name_data","metadata":[{"metadata":{"selected":true,"inclusion":"automatic","breadcrumb":["properties","name"]}},{"metadata":{"selected":true,"inclusion":"automatic","breadcrumb":["properties","name"]}},{"metadata":{"selected":true,"inclusion":"automatic","breadcrumb":["properties","name_type"]}},{"metadata":{"selected":true,"inclusion":"automatic","breadcrumb":["properties","name_type"]}},{"metadata":{"selected":true,"replication-method":"INCREMENTAL","inclusion":"available","valid-replication-keys":["name","name_type"],"table-key-properties":["name","name_type"],"breadcrumb":[]}}],"key_properties":["name","name_type"],"bookmark_properties":["name","name_type"]}
PlanetScale Tap : INFO : Stream "name_data" will be synced incrementally
PlanetScale Tap : INFO : known position is CgEtEg9sYXJnZS1pbXBvcnQtNTY=
PlanetScale Tap : INFO : syncing rows from stream "name_data" from shard "-" at position []
PlanetScale Tap : INFO : [name_data shard : -] peeking to see if there's any new rows
PlanetScale Tap : INFO : [name_data shard : -] syncing rows with cursor [shard:"-"  keyspace:"large-import-56"]
PlanetScale Tap : INFO : [name_data shard : -] latest database position is [MySQL56/094dfc93-c7f4-11ed-85a0-1af10d8c33d9:1-3676,0952eb04-c7f4-11ed-9488-9a951227f355:1-1175,8bf226de-ae66-11ed-8a7f-8e9fa0edf0b2:1-403,8c4d148f-ae66-11ed-a066-bed353af0f03:1-852,8d08e564-d884-11ed-b9fc-92807b25c579:1-2223,a6f10029-d7c4-11ed-80c4-3aff882caf60:1-1607,a9a3719d-d7c4-11ed-beb2-4a7ff2069748:1-12434,f4c25a4b-a7d9-11ed-86d1-8e2dfefa1b03:1-1759,f5c41ba5-a7d9-11ed-8902-369ef0b7d59c:1-4380]
PlanetScale Tap : INFO : flushing [1701] messages for stream "name_data" in [2] batches
PlanetScale Tap : INFO : Last row for table [name_data] in the last batch  has primary keys : [ { name : VARCHAR("Zane") } { name_type : VARCHAR("MALE_FIRST") } ]
PlanetScale Tap : INFO : [name_data shard : -] Finished reading all rows for table [name_data]
PlanetScale Tap : INFO : saving state to path : state/state-1684946956870.json
PlanetScale Tap : INFO : saving state to path : state/state-1684946956871.json
PlanetScale Tap : INFO : saving state to path : state/state-1684946956872.json
```

The above output shows that we've solved a few problems. 
1. Correctly order when we write records and when we write rows. 
``` bash
PlanetScale Tap : INFO : flushing [1701] messages for stream "name_data" in [2] batches
PlanetScale Tap : INFO : Last row for table [name_data] in the last batch  has primary keys : [ { name : VARCHAR("Zane") } { name_type : VARCHAR("MALE_FIRST") } ]
PlanetScale Tap : INFO : [name_data shard : -] Finished reading all rows for table [name_data]
PlanetScale Tap : INFO : saving state to path : state/state-1684946956870.json
```

2. Clearly identify which row was last sent to Stitch.
```
 Last row for table [name_data] in the last batch  has primary keys : [ { name : VARCHAR("Zane") } { name_type : VARCHAR("MALE_FIRST") } ]
```

which matches what stitch says
``` sql
mysql> select * from name_data order by _sdc_batched_at DESC LIMIT 1 \G
*************************** 1. row ***************************
         name_type: MALE_FIRST
              name: Zane
  _sdc_received_at: 2023-05-24 15:40:22.000
     _sdc_sequence: 1684942760572
   _sdc_batched_at: 2023-05-24 15:40:42.579
_sdc_table_version: 0
```

3. Since we're not stdout --> stdin for sending records around, we use an in-process list and serialize that ASAP, instead of waiting for the 4096 byte buffer to fill up and wait for the kernel to flush to stdin.

### Reference
[stdio buffering](https://www.pixelbeat.org/programming/stdio_buffering/)

### CODE REVIEW

File: cmd/http-tap/main.go [hunk 1]
```diff
@@ -65,7 +65,7 @@ func execute(logger internal.Logger, apiUrl string, batchSize, bufferSize int, t
 		if s != nil {
 			if s != stream && stream != nil {
 				// The schema we're writing out has changed, flush the messages so far.
-				if err := batchWriter.Flush(stream); err != nil {
+				if batchWriter.Flush(*stream); err != nil {
```
Comment (mdlayher) [inline cmd/http-tap/main.go:68]: This change is incorrect, the error needs to be captured in a var.

File: cmd/internal/http_record_writer.go [hunk 2]
```diff
@@ -15,49 +19,67 @@ const (
...
+
+	stateFileContents, err := json.Marshal(state)
+	if err != nil {
+		return err
+	}
+
+	statePath := filepath.Join(h.stateFileDir, fmt.Sprintf("state-%v.json", now.UnixMilli()))
+	h.logger.Info(fmt.Sprintf("saving state to path : %v", statePath))
+
+	if err := ioutil.WriteFile(statePath, stateFileContents, fs.ModePerm); err != nil {
```
Comment (mdlayher) [inline cmd/internal/http_record_writer.go:64]: Should use `os.WriteFile` in modern code.

File: cmd/internal/http_record_writer.go [hunk 3]
```diff
@@ -15,49 +19,67 @@ const (
...
+		h.logger.Info(fmt.Sprintf("unable to save state to path %v", statePath))
```
Comment (mdlayher) [inline cmd/internal/http_record_writer.go:65]: Error log level?

File: cmd/internal/planetscale_edge_database.go [hunk 4]
```diff
@@ -21,15 +21,18 @@ import (
...
 )
 
-var (
-	BinlogsPurgedMessage = "Cannot replicate because the master purged required binary logs"
+type (
+	OnResult func(*sqltypes.Result) error
+	OnCursor func(*psdbconnect.TableCursor) error
 )
 
+var BinlogsPurgedMessage = "Cannot replicate because the master purged required binary logs"
```
Comment (mdlayher) [inline cmd/internal/planetscale_edge_database.go:29]: It's odd that this is exported, can you unexport it?

File: cmd/internal/planetscale_edge_database.go [hunk 5]
```diff
@@ -21,15 +21,18 @@ import (
...
+
 // PlanetScaleDatabase is a general purpose interface
 // that defines all the data access methods needed for the PlanetScale Singer Tap to function.
 type PlanetScaleDatabase interface {
 	CanConnect(ctx context.Context, ps PlanetScaleSource) error
-	Read(ctx context.Context, ps PlanetScaleSource, s Stream, tc *psdbconnect.TableCursor, indexRows bool, columns []string) (*SerializedCursor, error)
+	Read(ctx context.Context, ps PlanetScaleSource, table Stream, lastKnownPosition *psdbconnect.TableCursor, columns []string, onResult OnResult, onCursor OnCursor) (*SerializedCursor, error)
```
Comment (mdlayher) [inline cmd/internal/planetscale_edge_database.go:35]: This has enough arguments to merit a struct IMO:

```
Read(ctx, ReadState{
    // ...
})
```

File: cmd/internal/planetscale_edge_database.go [hunk 6]
```diff
@@ -128,7 +129,7 @@ func (p PlanetScaleEdgeDatabase) Read(ctx context.Context, ps PlanetScaleSource,
 	}
 }
 
-func (p PlanetScaleEdgeDatabase) sync(ctx context.Context, tc *psdbconnect.TableCursor, stopPosition string, s Stream, columns []string, ps PlanetScaleSource, tabletType psdbconnect.TabletType, readDuration time.Duration, indexRows bool) (*psdbconnect.TableCursor, error) {
+func (p PlanetScaleEdgeDatabase) sync(ctx context.Context, tc *psdbconnect.TableCursor, stopPosition string, s Stream, columns []string, ps PlanetScaleSource, tabletType psdbconnect.TabletType, readDuration time.Duration, onResult OnResult, onCursor OnCursor) (*psdbconnect.TableCursor, error) {
```
Comment (mdlayher) [inline cmd/internal/planetscale_edge_database.go:132]: This also could really use a struct.

File: cmd/internal/planetscale_edge_database_test.go [hunk 7]
```diff
@@ -426,8 +432,8 @@ func TestRead_CanStopAtWellKnownCursor(t *testing.T) {
 
 	logLines := tal.logMessages
 	assert.Equal(t, "[customers shard : -] Finished reading all rows for table [customers]", logLines[len(logLines)-1])
-	records := tal.records["customers"]
-	assert.Equal(t, 2*(nextVGtidPosition/3), len(records))
+	// records := tal.records["customers"]
```
Comment (mdlayher) [inline cmd/internal/planetscale_edge_database_test.go:435]: remove?

File: cmd/internal/planetscale_edge_mysql.go [hunk 8]
```diff
@@ -211,6 +211,13 @@ func getJsonSchemaType(mysqlType string) StreamProperty {
 		return StreamProperty{Types: []string{"null", "string"}, CustomFormat: "timestamp_with_timezone"}
 	}
 
+	if strings.HasSuffix(mysqlType, "int") {
```
Comment (mdlayher) [inline cmd/internal/planetscale_edge_mysql.go:214]: Can you rely on this format for all integer types? Thinking of the Go convention of like `int32` vs strings like `bigint`.

File: cmd/internal/sync.go [hunk 9]
```diff
@@ -60,30 +63,81 @@ func Sync(ctx context.Context, mysqlDatabase PlanetScaleEdgeMysqlAccess, edgeDat
 		}
 
 		for shard, cursor := range streamShardStates {
-			logger.Info(fmt.Sprintf("syncing rows from stream %q from shard %q", stream.Name, shard))
+			defer func() {
```
Comment (mdlayher) [inline cmd/internal/sync.go:66]: Defers are function scoped and not loop scoped, so this seems suspect.

File: cmd/internal/sync.go [hunk 10]
```diff
@@ -60,30 +63,81 @@ func Sync(ctx context.Context, mysqlDatabase PlanetScaleEdgeMysqlAccess, edgeDat
...
+				recordWriter.Flush(stream)
+				return nil
+			}
+
+			newCursor, err := edgeDatabase.Read(ctx, source, stream, tc, stream.Metadata.GetSelectedProperties(), onResult, onCursor)
 			if err != nil {
 				return err
 			}
 
+			recordWriter.Flush(stream)
```
Comment (mdlayher) [inline cmd/internal/sync.go:100]: Check flush error.

File: cmd/internal/planetscale_edge_mysql.go (hunk 8 above)
Comment (Phanatic) [reply cmd/internal/planetscale_edge_mysql.go:214]: Yeah, `bigint` is captured above. 

From Mysql's list here : https://dev.mysql.com/doc/refman/8.0/en/integer-types.html

Comment (Phanatic) [reply cmd/internal/planetscale_edge_mysql.go:214]: I'll do another pass on the data types and make sure we do the right thing for Stitch here.

**Review by notfelineit — APPROVED** [review]
these interfaces have really come along 🥳

**Pull request merged**